File: 1299-replace-elements-with-greatest-element-on-right-side/1299-replace-elements-with-greatest-element-on-right-side.cpp
class Solution {
public:
    vector<int> replaceElements(vector<int>& arr) {
        vector<int> res(arr.size(),-1);
        stack<int> st;
        
        for(int i=arr.size()-1; i>=0; i--){
            if(st.size() > 0){
                res[i] = st.top();
                if(arr[i] > st.top()){
                    st.pop();
                    st.push(arr[i]);
                }
            }   
            else{
                st.push(arr[i]);
            }
        }
        return res;
    }
};

File: 1299-replace-elements-with-greatest-element-on-right-side/1299-replace-elements-with-greatest-element-on-right-side.py
class Solution:
    def replaceElements(self, arr: List[int]) -> List[int]:
        st = []
        
        res = [-1]*len(arr)
        
        for i in range(len(arr)-1,-1,-1):
            if st:
                res[i] = st[-1]
                if arr[i] > st[-1]:
                    st.pop()
                    st.append(arr[i])
            else:
                st.append(arr[i])
            
        return res